### PR TITLE
Migrate reportAction.originalMessage to transactions

### PR DESCRIPTION
**Pull request by luacmartins**
### Details
Migrate instances of `action.originalMessage.<amount | currency | comment | etc>` to `transaction.<>` wherever possible

### Fixed Issues
$ https://github.com/Expensify/App/issues/22208

### Tests
1. Request money from someone multiple times `+ > Request money`
2. Make sure to request money with and without comments
3. Verify that the report preview shows the correct data (amount, currency, comment)
4. Tap on the report preview
5. Verify the the IOU Preview shows the correct data
6. Tap the IOU preview
7. Verify the the request details in the thread show the correct data
8. Delete the request
9. Verify the the request disappears 
10. Repeat the process for `Split bill`
11. Tap the preview and verify that the details shown are correct

- [x] Verify that no errors appear in the JS console

### Offline tests
N/A

### QA Steps
Same as tests

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

web

</details>

<details>
<summary>Mobile Web - Chrome</summary>

chrome

</details>

<details>
<summary>Mobile Web - Safari</summary>

safari

</details>

<details>
<summary>Desktop</summary>

desktop

</details>

<details>
<summary>iOS</summary>

ios

</details>

<details>
<summary>Android</summary>

android

</details>

### CODE REVIEW

File: src/libs/ReportActionsUtils.js [hunk 1]
```diff
@@ -138,6 +150,17 @@ function getParentReportActionInReport(childReportID, parentReportID) {
...
+ * Get the details linked to the IOU reportAction
+ *
+ * @param {Object} reportAction
+ * @returns {Object}
+ */
+function getLinkedTransaction(reportAction = {}) {
+    const transactionID = lodashGet(reportAction, ['originalMessage', 'IOUTransactionID'], '');
+    return allTransactions[`${ONYXKEYS.COLLECTION.TRANSACTION}${transactionID}`] || {};
+}
+
```
Comment (mountiny) [inline src/libs/ReportActionsUtils.js:163]: Should we move this to transactionUtils? We already have the allTransactions there

File: src/libs/ReportUtils.js [hunk 2]
```diff
@@ -1250,6 +1224,33 @@ function getMoneyRequestReportName(report, policy = undefined) {
...
+
+/**
+ * Gets transaction amount, currency and comment
+ *
+ * @param {Object} transaction
+ * @returns {Object}
+ */
+function getTransactionDetails(transaction) {
+    const reportID = transaction.reportID;
+    const report = getReport(reportID);
```
Comment (mountiny) [inline src/libs/ReportUtils.js:1245]: Its not reused so I think we dont need to assign it to new variable
```suggestion
    const report = getReport(transaction.reportID);
```

File: src/libs/ReportUtils.js [hunk 3]
```diff
@@ -1250,6 +1224,33 @@ function getMoneyRequestReportName(report, policy = undefined) {
...
+ *
+ * @param {String} reportID
+ * @returns {Object}
+ */
+function getReport(reportID) {
+    return lodashGet(allReports, `${ONYXKEYS.COLLECTION.REPORT}${reportID}`, {});
+}
+
+/**
+ * Gets transaction amount, currency and comment
```
Comment (mountiny) [inline src/libs/ReportUtils.js:1238]:
```suggestion
 * Gets transaction amount, created, currency and comment values
```

File: src/libs/ReportUtils.js [hunk 4]
```diff
@@ -1250,6 +1224,33 @@ function getMoneyRequestReportName(report, policy = undefined) {
     return payerPaidAmountMesssage;
 }
 
+/**
+ * Get the report for a reportID
```
Comment (mountiny) [inline src/libs/ReportUtils.js:1228]:
```suggestion
 * Get the report given a reportID
```

**Comment by luacmartins** [comment]
This PR is ready for review. I found a couple of unrelated bugs while testing:

1. `<MoneyRequestView>` doesn't update with the edited amounts. This happens because [this memo](https://github.com/Expensify/App/blob/add8d3cc0a05259297c5d16ac4921a47dc367963/src/pages/home/report/ReportActionItem.js#L623-L648) prevents the parent from re-rendering so props never change and we also don't store the transaction in state so the component never re-renders
2. We're creating optimistic transactions for expense reports with positive values [here](https://github.com/Expensify/App/blob/add8d3cc0a05259297c5d16ac4921a47dc367963/src/libs/actions/IOU.js#L367), which leads to IOUPreviews displaying a negative amount initially (until the page is refreshed

**Comment by mountiny** [comment]
Thanks Carlos! 

> We're creating optimistic transactions for expense reports with positive values [here](https://github.com/Expensify/App/blob/add8d3cc0a05259297c5d16ac4921a47dc367963/src/libs/actions/IOU.js#L367), which leads to IOUPreviews displaying a negative amount initially (until the page is refreshed)

I think we should do this one in this PR

**Comment by mountiny** [comment]
@luacmartins Actually made the PR here and also assigned @sobitneupane to it https://github.com/Expensify/App/pull/24613

Noticed one unnecessary use of the transaction too and removed it there.

**Review by sobitneupane — COMMENTED** [review]
Report Preview is not showing the amount in previous transactions (especially split bill).  

https://github.com/Expensify/App/assets/25876548/b75e0e70-664c-4442-8875-05c317f98530


Also, after creating a new transaction (for example: split bill), report preview at first shows 0 and after a split second, changes to actual value.


https://github.com/Expensify/App/assets/25876548/2efbcc36-26d4-4d52-82ef-d08a0b5d477f

**Comment by mountiny** [comment]
@luacmartins should this be onyx.set? https://github.com/Expensify/App/blob/b35fa79f6f1aafa3598e6776c87fa6561b9e6a21/src/libs/actions/IOU.js#L527-L531 to make the effect show immediately

**Comment by mountiny** [comment]
Some merge conflicts appeared

**Comment by luacmartins** [comment]
>Report Preview is not showing the amount in previous transactions (especially split bill).

This is a backend issue and shouldn't be a blocker for this PR. The issue happens because we only return the `transactions` key for IOU/Expense reports, but the Split action is on a Chat report. I'll work on a fix for it in the backend.

>Also, after creating a new transaction (for example: split bill), report preview at first shows 0 and after a split second, changes to actual value.

This has been fixed.

**Comment by mountiny** [comment]
@sobitneupane Are you able to take another look?

**Review by sobitneupane — COMMENTED** [review]
### Screenshots/Videos
<details>
<summary>Web</summary>


https://github.com/Expensify/App/assets/25876548/ebc53595-b63b-4b8e-9703-3e845937fb54

https://github.com/Expensify/App/assets/25876548/31639f5b-00ac-42a7-bccd-d3cb2fd956c8


https://github.com/Expensify/App/assets/25876548/4b9a467e-5a30-46da-b8cb-de0fb7ad0fb6


</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/25876548/27ddc27f-4f5b-49ad-936c-ac36dfc5303f

</details>

<details>
<summary>Mobile Web - Safari</summary>


Screenshot 2023-08-17 at 00 06 37


</details>



<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/25876548/7e29f0cb-e803-4a45-88f6-40154f2c3e25

</details>

<details>
<summary>iOS</summary>



https://github.com/Expensify/App/assets/25876548/02802540-d9ed-42fd-83f2-10200485c7b1



</details>

<details>
<summary>Android</summary>






</details>

**Review by sobitneupane — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

File: src/libs/actions/IOU.js [hunk 5]
```diff
@@ -939,9 +938,9 @@ function deleteMoneyRequest(transactionID, reportAction, isSingleTransactionView
             updatedIOUReport = {...iouReport};
 
             // Because of the Expense reports are stored as negative values, we add the total from the amount
-            updatedIOUReport.total += reportAction.originalMessage.amount;
+            updatedIOUReport.total += transaction.amount;
```
Comment (mountiny) [inline src/libs/actions/IOU.js:941]: Should we use the getters here too

File: src/libs/actions/IOU.js [hunk 6]
```diff
@@ -939,9 +938,9 @@ function deleteMoneyRequest(transactionID, reportAction, isSingleTransactionView
...
         } else {
-            updatedIOUReport = IOUUtils.updateIOUOwnerAndTotal(iouReport, reportAction.actorAccountID, reportAction.originalMessage.amount, reportAction.originalMessage.currency, true);
+            updatedIOUReport = IOUUtils.updateIOUOwnerAndTotal(iouReport, reportAction.actorAccountID, transaction.amount, transaction.currency, true);
```
Comment (mountiny) [inline src/libs/actions/IOU.js:943]: Here as well

**Review by mountiny — COMMENTED** [review]
Two notes

**Review by mountiny — APPROVED** [review]
Thanks @sobitneupane and @luacmartins

File: src/libs/ReportUtils.js [hunk 7]
```diff
@@ -1229,6 +1203,32 @@ function getMoneyRequestReportName(report, policy = undefined) {
...
 }
 
+/**
+ * Get the report given a reportID
+ *
+ * @param {String} reportID
+ * @returns {Object}
+ */
+function getReport(reportID) {
+    return lodashGet(allReports, `${ONYXKEYS.COLLECTION.REPORT}${reportID}`, {});
```
Comment (aldo-expensify) [inline src/libs/ReportUtils.js:1213]: Why are we returning {} for missing reports? 😬 wouldn't `null` or `undefined` be better in these cases?

File: src/libs/ReportUtils.js [hunk 8]
```diff
@@ -1368,6 +1371,19 @@ function getModifiedExpenseOriginalMessage(oldTransaction, transactionChanges, i
...
 
+/**
+ * Returns the parentReport if the given report is a thread.
+ *
+ * @param {Object} report
+ * @returns {Object}
+ */
+function getParentReport(report) {
+    if (!report || !report.parentReportID) {
+        return {};
```
Comment (aldo-expensify) [inline src/libs/ReportUtils.js:1382]: Same as https://github.com/Expensify/App/pull/24345/files#r1335355086

File: src/libs/ReportUtils.js (hunk 7 above)
Comment (aldo-expensify) [reply src/libs/ReportUtils.js:1213]: It is much more intuitive for me to check if a report is missing with `report === null` than `Object.keys(report) === 0`

Comment (aldo-expensify) [reply src/libs/ReportUtils.js:1213]: oh, I see that you just moved around this function and it was already like that 😿

**Pull request merged**